contributors/design-proposals/node/node-fence.md

```diff
@@ -0,0 +1,528 @@
+# Node fencing
+Status: Pending
+
+Version: Alpha
+
+Implementation Owner: Yaniv Bronhaim (ybronhei@redhat.com)
+
+## Overview
+Rationale and architecture for the addition of a fencing mechanism in Kubernetes clusters.
+
+### Motivation
+In Kubernetes a cluster can get into a network partition state between nodes to the api server running on the master node (e.g. network device failure). When that happens their nodes’ status is changed to “not ready” by the node controller and the scheduler cannot know the status of pods running on that node without reaching the nodes’ kubelet service. From such scenario, k8s (since 1.8) defines an eviction timeout (forced by the node controller) such that after 5 minutes pods will enter a termination state in the api server and will try to shutdown gracefully when possible (this happens only if the connectivity returns).
+
+If the pod belongs to a ReplicaSet, once the termination state is set, the ReplicaSet controller will immediately start another instance of that pod.  However when the pod is part of a StatefulSet, the StatefulSet controller won’t start new instance of that pod until the kubelet responds with the pod’s status. This is because there is a permanent association between a StatefulSet member and its storage.  Not waiting would potentially result in multiple copies of a member, all writing to the same volume (leading to corruption or worse).
+
+### Cloud deployments
+When k8s is deployed on a cloud such as AWS or GCE, the autoscaler uses the Cloud Provider APIs to remove failed nodes (Failure detection and allowing recovery of StatefulSets is more of a side effect of the autoscaler which is more focused on using the Cloud Provider APIs to monitor load and trigger scale up/down events), effectively bounding the amount of time the node will stay in the “not ready” state and how long the scheduler will need to wait before it can safely start the pod elsewhere.  However in the case of bare metal, no such mechanism exists and recovery will be blocked until an admin intervenes.
+
+This is particularly problematic because all StatefulSet scaling events (up and down) block until existing failures have been recovered.  From some perspectives, it could be considered that each node hosting a StatefulSet member has become a single point of failure for that set.
+
+### Scope of this proposal
+Managing HA entities in cluster can relate to nodes and applications. In this scope we cover “node isolation” actions while node is not responsive until it becomes responsive again. Node isolation can be done by power management action (rebooting the node’s machine), storage fencing (preventing node from accessing), and cluster-wide isolation actions.
```

Comment:
I don't understand the first sentence.

Reply:
I wanted to refer to HA applications and mixed it up during the sentence. I'll remove this first sentence